include/seqan3/range/view/async_buffer.hpp

```diff
@@ -0,0 +1,538 @@
+                    #ifndef NDEBUG
+                        has_value = true;
+                    #endif
+                        return *this;
+                    default:
+                        assert(s == contrib::queue_op_status::empty);
+                        break; // loop again
+                }
+
+                delay.wait(); // pause and then try again.
+            }
+
+            // TODO [[unreachable]]
+            return *this;
+        }
+
+        //!\brief Post-increment.
+        void operator++(int) noexcept
+        {
+            ++(*this);
+        }
+        //!\}
+
+        /*!\name Comparison operators
+        * \{
+        */
+        //!\brief Compares for equality with sentinel.
+        friend constexpr bool operator==(async_buffer_iterator const & lhs,
+                                         std::ranges::default_sentinel_t const &) noexcept
+        {
+            return lhs.at_end;
+        }
+
+        //!\copydoc operator==
+        friend constexpr bool operator==(std::ranges::default_sentinel_t const &,
+                                         async_buffer_iterator const & rhs) noexcept
+        {
+            return rhs == std::ranges::default_sentinel_t{};
+        }
+
+        //!\brief Compares for inequality with sentinel.
+        friend constexpr bool operator!=(async_buffer_iterator const & lhs,
+                                         std::ranges::default_sentinel_t const &) noexcept
+        {
+            return !(lhs == std::ranges::default_sentinel_t{});
+        }
+
+        //!\copydoc operator!=
+        friend constexpr bool operator!=(std::ranges::default_sentinel_t const &,
+                                         async_buffer_iterator const & rhs) noexcept
+        {
+            return rhs != std::ranges::default_sentinel_t{};
+        }
+        //!\}
+    };
+
+public:
+    /*!\name Constructor, destructor, and assignment.
+     * \{
+     */
+    async_buffer_view()                                         = default; //!< Defaulted.
+    async_buffer_view(async_buffer_view const &)                = delete;  //!< Deleted, because not a forward range.
+    async_buffer_view(async_buffer_view &&)                     = default; //!< Defaulted.
+    async_buffer_view & operator=(async_buffer_view const &)    = delete;  //!< Deleted, because not a forward range.
+    async_buffer_view & operator=(async_buffer_view &&)         = default; //!< Defaulted.
+
+    /*!\brief Tell thread to stop and wait for it.
+     *
+     * \details
+     *
+     * Destruction of the view clears the buffer and temporary storage. Up to `n + 1` elements may be destructed
+     * together with the view. [This may be important if the underlying range is also single-pass and you
+     * are trying to estimate how many elements will be consumed – in case you do not plan on consuming the
+     * entire range.]
+     */
+    ~async_buffer_view()
+    {
+        stop_producer.store(true);
```

Comment:
You also need to wait until the queue is empty after it was closed, otherwise you get a data race with destruction of the queue and consumer threads. If not you would need to add some inter-thread signaling, which is kind of complex in this model.